### DISCUSSION TITLE
Write atom and async behavior

### DISCUSSION THREAD

**Discussion by chikunov** [pmndrs/jotai#1909] [Q&A] [answered]
Hello everyone!

I'm trying to see if I understand the behavior correctly in the following case:
```js
const sharedDataAtom = atom({
 ...
});

const someWriteAtom = atom(null, async (get, set, id: string) => {
  const sharedData = get(sharedDataAtom);
  const backendData = await backendRequest(sharedData.id);
  const updatedData = merge(sharedData, backendData); // sharedData here might be already outdated due to await above.
  set(sharedDataAtom, updatedData); // this is going to override sharedAtom value that was updated meanwhile
});

const myFunc = useSetAtom(someWriteAtom);
useEffect(() => {
  myFunc(1);
  myFunc(2);
  myFunc(3);
}, []);
```

basically, having multiple requests in parallel here would lead to a corrupted data because every setter would just independently override results. Is this true? And it is true that something like this would solve it:
```js
const someWriteAtom = atom(null, async (get, set, id: string) => {
  const sharedData = get(sharedDataAtom);
  const backendData = await backendRequest(sharedData.id);
  const updatedData = merge(get(sharedDataAtom), backendData); // get sharedData again to have it up-to-date.
  set(sharedDataAtom, updatedData);
});
```

**Comment by himself65** [comment]
I think Atom will always accept the last update. So for the async function, which might have timing issues. I suggest using `switchMap` from `rxjs` to always accept the last update

**Comment by dai-shi** [comment, marked as answer]
> And it is true that something like this would solve it:
> 
> ```js
> const someWriteAtom = atom(null, async (get, set, id: string) => {
>   const sharedData = get(sharedDataAtom);
>   const backendData = await backendRequest(sharedData.id);
>   const updatedData = merge(get(sharedDataAtom), backendData); // get sharedData again to have it up-to-date.
>   set(sharedDataAtom, updatedData);
> });
> ```

Yes.
